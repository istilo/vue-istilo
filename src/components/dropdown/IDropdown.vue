<script setup lang="ts">
import { computed, onBeforeMount, onUnmounted, ref } from 'vue';

const emit = defineEmits(['close']);
const dropdown = ref();

const props = withDefaults(
	defineProps<{
		/**
		 * Toggles the dropdown menu
		 */
		isOpen: boolean;

		/**
		 * Specify which side the dropdown will show
		 */
		position?: string;
	}>(),
	{
		isOpen: false,
	}
);

onBeforeMount(() => {
	window.addEventListener('click', close);
});

onUnmounted(() => {
	window.removeEventListener('click', close);
});

const getMenuClasses = computed(() => {
	return [
		'i-dropdown',
		props.isOpen ? 'i-open' : '',
		props.position ? `i-dropdown-${props.position}` : '',
	];
});

const close = (event: { target: any }) => {
	if (props.isOpen && !dropdown.value.contains(event.target)) {
		emit('close');
	}
};
</script>

<template>
	<div
		class="i-inline"
		ref="dropdown">
		<slot />

		<div :class="getMenuClasses">
			<slot name="menu" />
		</div>
	</div>
</template>
