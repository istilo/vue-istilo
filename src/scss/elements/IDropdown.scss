@import '@/scss/helpers';
@import '@/scss/functions';
@import '@/scss/mixins';

// ============================================================
// Name:            Dropdown
// Description:     Component to create dropdown menus
//
// Component:       `#{$prefix}dropdown`
//
// Adopted:         `#{$prefix}dropdown-nav`
//
// Modifiers:       `#{$prefix}dropdown-top-*`
//                  `#{$prefix}dropdown-bottom-*`
//                  `#{$prefix}dropdown-left-*`
//                  `#{$prefix}dropdown-right-*`
//                  `#{$prefix}dropdown-stack`
//                  `#{$prefix}dropdown-grid`
//
// ============================================================

// Variables
// ============================================================

$dropdown-z-index: layer(dropdown) !default;
$dropdown-min-width: 200px !default;
$dropdown-max-width: 300px !default;
$dropdown-padding: gutter(small) - 5px !default;
$dropdown-background: color(inverse) !default;
$dropdown-color: color(default) !default;
$dropdown-margin: margin(small) !default;
$dropdown-offset-top: margin(medium) + 10px !default;

$dropdown-nav-item-color: color(muted) !default;
$dropdown-nav-item-hover-color: color(default) !default;
$dropdown-nav-header-color: color(emphasis) !default;

$dropdown-nav-divider-border-width: border(width) !default;
$dropdown-nav-divider-border: border(color) !default;
$dropdown-nav-sublist-item-color: color(muted) !default;
$dropdown-nav-sublist-item-hover-color: color(default) !default;

$dropdown-box-shadow: box-shadow(medium) !default;

/* ============================================================
    Component: Dropdown
   ============================================================ */

/*
 * 1. Hide by default
 * 2. Set position
 * 3. Set a default width
 * 4. Style
 */

.#{$prefix}dropdown {
	/* 1 */
	display: none;

	/* 2 */
	position: absolute;

	top: $dropdown-offset-top;
	z-index: $dropdown-z-index;

	/* 3 */
	box-sizing: border-box;
	min-width: $dropdown-min-width;
	max-width: $dropdown-max-width;

	/* 4 */
	padding: $dropdown-padding;
	background: $dropdown-background;
	color: $dropdown-color;
	box-shadow: $dropdown-box-shadow;
}

/* Responsive Dropdown
   ============================================================ */

/*
 * 1. We will fixed the position of dropdown on bottom of the mobile
 * 2. We use transform to animate smooth by sliding up
 * 3. Remove the margin by default
 * 4. We added 100 more on layer to overlap on lightbox
 * 5. Style
 */
@include respond-to-max(small) {
	.#{$prefix}dropdown {
		/* 1 */
		position: fixed;
		top: auto;
		bottom: 0;
		left: 0;
		right: 0;

		/* 2 */
		transform: translateY(100%);

		/* 3 */
		margin-left: 0px !important;

		/* 4 */
		z-index: layer(lightbox) + 100;

		/* 5 */
		width: 100%;
		max-width: 100%;
		transition: all 0.3s ease-in-out;
	}

	.#{$prefix}dropdown.#{$prefix}slideup {
		transform: translateY(0%);
	}

	body.#{$prefix}dropdown-open:before {
		content: '';
		background: #000000;
		opacity: 0;
		position: fixed;
		height: 100%;
		width: 100%;
		left: 0;
		box-sizing: border-box;

		z-index: layer(lightbox);
		overflow-y: hidden;

		transition: all 0.3s ease-in-out;
	}

	body.#{$prefix}dropdown-open.#{$prefix}open:before {
		opacity: 0.3;
	}

	html.#{$prefix}dropdown-open-container {
		overflow: hidden;
	}
}

/*
 * Use drop-up when the dropdown is on the bottom of the screen
 * to flip the position
 */
.#{$prefix}dropdown.#{$prefix}dropup {
	top: auto;
	bottom: 0 !important;
}

/* Show */
.#{$prefix}dropdown.#{$prefix}open {
	display: block;
}

/* Nav
 * Adopts `#{$prefix}nav`
   ============================================================ */

.#{$prefix}dropdown-nav {
	white-space: nowrap;
}

/*
 * Items
 */

.#{$prefix}dropdown-nav > li > a {
	color: $dropdown-nav-item-color;
	text-decoration: none;
}

/* Hover + Focus + Active */
.#{$prefix}dropdown-nav > li > a:hover,
.#{$prefix}dropdown-nav > li > a:focus,
.#{$prefix}dropdown-nav > li.#{$prefix}active > a {
	color: $dropdown-nav-item-hover-color;
	text-decoration: none;
}

/*
 * Header
 */

.#{$prefix}dropdown-nav .#{$prefix}nav-header {
	color: $dropdown-nav-header-color;
}

/*
 * Divider
 */

.#{$prefix}dropdown-nav .#{$prefix}nav-divider {
	border-top: $dropdown-nav-divider-border-width solid $dropdown-nav-divider-border;
}

/*
 * Sublists
 */

.#{$prefix}dropdown-nav .#{$prefix}nav-sub a {
	color: $dropdown-nav-sublist-item-hover-color !important;
}

/* Hover + Focus */
.#{$prefix}dropdown-nav .#{$prefix}nav-sub a:hover,
.#{$prefix}dropdown-nav .#{$prefix}nav-sub a:focus {
	color: $dropdown-nav-sublist-item-hover-color !important;
}

/* Direction / Alignment modifiers
   ============================================================ */

/* Direction */
[class*='#{$prefix}dropdown-top'] {
	margin-top: (-$dropdown-margin);
}

[class*='#{$prefix}dropdown-bottom'] {
	margin-top: $dropdown-margin;
}

[class*='#{$prefix}dropdown-left'] {
	margin-left: (-$dropdown-margin);
}

[class*='#{$prefix}dropdown-right'] {
	margin-left: $dropdown-margin;
}

.#{$prefix}dropdown-right {
	right: 0;
}

.#{$prefix}dropdown-left {
	left: 0;
}

/* Grid modifiers
   ============================================================ */

.#{$prefix}dropdown-stack .#{$prefix}dropdown-grid > * {
	width: 100% !important;
}
