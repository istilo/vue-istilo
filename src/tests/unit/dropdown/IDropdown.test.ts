import { describe, it, expect, beforeEach, afterEach } from 'vitest';
import { shallowMount, VueWrapper, enableAutoUnmount } from '@vue/test-utils';

import IDropdown from '@/components/dropdown/IDropdown.vue';

enableAutoUnmount(afterEach);

describe('TAS Dropdown', () => {
	let wrapper: VueWrapper<any>;

	beforeEach(() => {
		wrapper = shallowMount(IDropdown, {
			propsData: {
				isOpen: false,
			},
			slots: {},
		});

		expect(wrapper.exists()).toBe(true);
	});

	afterEach(() => {
		expect(wrapper.html()).toMatchSnapshot();
	});

	it('Shoulder set the default element and classes', () => {
		expect(wrapper.classes()).toContain('i-inline');
		expect(wrapper.get('.i-dropdown')).toBeDefined();
	});

	it('Shoulder set the position class modifier', async () => {
		await wrapper.setProps({
			position: 'right',
		});

		expect(wrapper.get('.i-dropdown').classes()).toContain('i-dropdown-right');
	});

	it('Shoulder set the open class modifier', async () => {
		await wrapper.setProps({
			isOpen: true,
		});

		expect(wrapper.get('.i-open')).toBeDefined();
	});

	it('Shoulder have button and menu list as child components.', async () => {
		const button = '<button>Dropdown</button>';
		const menu = '<ul><li>Option 1</li><li>Option 2</li></ul>';

		wrapper = shallowMount(IDropdown, {
			propsData: {
				isOpen: false,
			},
			slots: {
				default: button,
				menu: menu,
			},
		});

		expect(wrapper.get('button')).toBeDefined();
		expect(wrapper.get('ul')).toBeDefined();
	});
});
