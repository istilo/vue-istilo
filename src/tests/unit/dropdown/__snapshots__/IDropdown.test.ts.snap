// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TAS Dropdown > Shoulder have button and menu list as child components. 1`] = `
"<div class=\\"i-inline\\"><button>Dropdown</button>
  <div class=\\"i-dropdown\\">
    <ul>
      <li>Option 1</li>
      <li>Option 2</li>
    </ul>
  </div>
</div>"
`;

exports[`TAS Dropdown > Shoulder set the default element and classes 1`] = `
"<div class=\\"i-inline\\">
  <div class=\\"i-dropdown\\"></div>
</div>"
`;

exports[`TAS Dropdown > Shoulder set the open class modifier 1`] = `
"<div class=\\"i-inline\\">
  <div class=\\"i-dropdown i-open\\"></div>
</div>"
`;

exports[`TAS Dropdown > Shoulder set the position class modifier 1`] = `
"<div class=\\"i-inline\\">
  <div class=\\"i-dropdown i-dropdown-right\\"></div>
</div>"
`;
